<a href="https://colab.research.google.com/github/drewamorbordelon/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS19_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [ ]:
### YOUR WORK HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data

--2020-08-11 20:17:24--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data’

imports-85.data     100%[===================>]  25.33K  --.-KB/s    in 0.03s   

2020-08-11 20:17:24 (876 KB/s) - ‘imports-85.data’ saved [25936/25936]



In [ ]:
import pandas as pd
import numpy as np

In [ ]:
column_headers = ['symboling','normalized','make','fuel-type', 'aspiration', 
                  'num-of-doors', 'body-style', 'drive-wheels', 
                  'engine-location', 'wheel-base','length', 'width', 'height',
                  'curb-weight', 'engine-type', 'num-of-cylinders',
                  'engine-size', 'fuel-system', 'bore', 'stroke', 'compression', 
                  'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

In [ ]:
df = pd.read_csv('/content/imports-85.data', names=column_headers) # sep=";"

In [ ]:
print(df.shape)
df.head()

(205, 26)


symboling normalized         make  ... city-mpg highway-mpg  price
0          3          ?  alfa-romero  ...       21          27  13495
1          3          ?  alfa-romero  ...       21          27  16500
2          1          ?  alfa-romero  ...       19          26  16500
3          2        164         audi  ...       24          30  13950
4          2        164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [ ]:
df.tail()

symboling normalized   make  ... city-mpg highway-mpg  price
200         -1         95  volvo  ...       23          28  16845
201         -1         95  volvo  ...       19          25  19045
202         -1         95  volvo  ...       18          23  21485
203         -1         95  volvo  ...       26          27  22470
204         -1         95  volvo  ...       19          25  22625

[5 rows x 26 columns]

In [ ]:
df.isnull().sum()

symboling           0
normalized          0
make                0
fuel-type           0
aspiration          0
num-of-doors        0
body-style          0
drive-wheels        0
engine-location     0
wheel-base          0
length              0
width               0
height              0
curb-weight         0
engine-type         0
num-of-cylinders    0
engine-size         0
fuel-system         0
bore                0
stroke              0
compression         0
horsepower          0
peak-rpm            0
city-mpg            0
highway-mpg         0
price               0
dtype: int64

In [ ]:
df.describe()

symboling  wheel-base      length  ...  compression    city-mpg  highway-mpg
count  205.000000  205.000000  205.000000  ...   205.000000  205.000000   205.000000
mean     0.834146   98.756585  174.049268  ...    10.142537   25.219512    30.751220
std      1.245307    6.021776   12.337289  ...     3.972040    6.542142     6.886443
min     -2.000000   86.600000  141.100000  ...     7.000000   13.000000    16.000000
25%      0.000000   94.500000  166.300000  ...     8.600000   19.000000    25.000000
50%      1.000000   97.000000  173.200000  ...     9.000000   24.000000    30.000000
75%      2.000000  102.400000  183.100000  ...     9.400000   30.000000    34.000000
max      3.000000  120.900000  208.100000  ...    23.000000   49.000000    54.000000

[8 rows x 10 columns]

In [ ]:
df.describe(exclude='number')

normalized    make fuel-type aspiration  ... stroke horsepower peak-rpm price
count         205     205       205        205  ...    205        205      205   205
unique         52      22         2          2  ...     37         60       24   187
top             ?  toyota       gas        std  ...   3.40         68     5500     ?
freq           41      32       185        168  ...     20         19       37     4

[4 rows x 16 columns]

In [ ]:
df = df.replace({'?': np.NaN})

df.isnull().sum()

symboling            0
normalized          41
make                 0
fuel-type            0
aspiration           0
num-of-doors         2
body-style           0
drive-wheels         0
engine-location      0
wheel-base           0
length               0
width                0
height               0
curb-weight          0
engine-type          0
num-of-cylinders     0
engine-size          0
fuel-system          0
bore                 4
stroke               4
compression          0
horsepower           2
peak-rpm             2
city-mpg             0
highway-mpg          0
price                4
dtype: int64

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estaimated sample mean to the population mean. Can you say that they are different? 

In [ ]:
### YOUR WORK HERE
#sample = df.sample(10, random_state= 100)

sample = df.sample(10, random_state= 30)

print(sample.shape)
sample

(10, 26)


symboling normalized           make  ... city-mpg highway-mpg  price
121          1        154       plymouth  ...       31          38   6692
176         -1         65         toyota  ...       27          32  10898
24           1        148          dodge  ...       31          38   6229
88          -1        137     mitsubishi  ...       23          30   9279
72           3        142  mercedes-benz  ...       16          18  35056
103          0        108         nissan  ...       19          25  13499
26           1        148          dodge  ...       31          38   7609
20           0         81      chevrolet  ...       38          43   6575
155          0         91         toyota  ...       27          32   8778
185          2         94     volkswagen  ...       27          34   8195

[10 rows x 26 columns]

In [ ]:
df.describe()

symboling  wheel-base      length  ...  compression    city-mpg  highway-mpg
count  205.000000  205.000000  205.000000  ...   205.000000  205.000000   205.000000
mean     0.834146   98.756585  174.049268  ...    10.142537   25.219512    30.751220
std      1.245307    6.021776   12.337289  ...     3.972040    6.542142     6.886443
min     -2.000000   86.600000  141.100000  ...     7.000000   13.000000    16.000000
25%      0.000000   94.500000  166.300000  ...     8.600000   19.000000    25.000000
50%      1.000000   97.000000  173.200000  ...     9.000000   24.000000    30.000000
75%      2.000000  102.400000  183.100000  ...     9.400000   30.000000    34.000000
max      3.000000  120.900000  208.100000  ...    23.000000   49.000000    54.000000

[8 rows x 10 columns]

In [ ]:
from scipy import stats

In [ ]:
sample.describe()

symboling  wheel-base      length  ...  compression   city-mpg  highway-mpg
count  10.000000   10.000000   10.000000  ...    10.000000  10.000000    10.000000
mean    0.600000   96.430000  169.500000  ...     8.930000  27.000000    32.800000
std     1.264911    2.959373    9.499474  ...     0.634298   6.411795     7.269418
min    -1.000000   93.700000  157.300000  ...     7.500000  16.000000    18.000000
25%     0.000000   93.900000  160.925000  ...     8.775000  24.000000    30.500000
50%     0.500000   96.000000  170.700000  ...     9.000000  27.000000    33.000000
75%     1.000000   97.125000  174.800000  ...     9.400000  31.000000    38.000000
max     3.000000  102.400000  184.600000  ...     9.600000  38.000000    43.000000

[8 rows x 10 columns]

In [ ]:
# Performing a t test 

stats.ttest_1samp(sample['highway-mpg'], df['highway-mpg'].mean())

Ttest_1sampResult(statistic=0.8912422391619923, pvalue=0.39600040771093326)

**The sample is reflective of the population due to a p value being greater that .05 (DO NOT reject the null hypothesis). 
There could still be a larger sample to get a more accurate representation of the population results.**

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [ ]:
### YOUR WORK HERE

sample = df.sample(10, random_state= 30)

stats.ttest_1samp(sample['highway-mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

**Yes, the p value verifies his claim. P value is greater than .05.**

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [ ]:
### YOUR WORK HERE

sample_100 = df.sample(100, random_state= 30)
stats.ttest_1samp(sample_100['highway-mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

**No, different conclusion.**


Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

**Since we are using a higher sample value of n=100, we are getting a more statistcally significant p-value due to the fact that it is a much larger sample.**   


## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [ ]:
### YOUR WORK HERE

!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-08-11 02:31:38--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data.2’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-08-11 02:31:39 (286 KB/s) - ‘house-votes-84.data.2’ saved [18171/18171]



In [ ]:
import pandas as pd
import numpy as np
from scipy import stats

In [ ]:
column_headers = ['Class Name', 'handicapped','water-project-cost-sharing', 
                  'adoption-of-budget-resolution','physician-free-freeze',
                  'el-salvador-aid','religious-groups-in-school',
                  'anti-satellite-test-ban','aid-to-nicaraguan-contras',
                  'mx-missile','immigration','synfuels-corporation-cutback',
                  'education-spending','superfund-right-to-sue','crime',
                  'duty-free-exports','export-administration-act-south-africa']

In [ ]:
df = pd.read_csv('/content/house-votes-84.data', names=column_headers)

df.head()

Class Name  ... export-administration-act-south-africa
0  republican  ...                                      y
1  republican  ...                                      ?
2    democrat  ...                                      n
3    democrat  ...                                      y
4    democrat  ...                                      y

[5 rows x 17 columns]

In [ ]:
df.isnull().sum()

Class Name                                0
handicapped                               0
water-project-cost-sharing                0
adoption-of-budget-resolution             0
physician-free-freeze                     0
el-salvador-aid                           0
religious-groups-in-school                0
anti-satellite-test-ban                   0
aid-to-nicaraguan-contras                 0
mx-missile                                0
immigration                               0
synfuels-corporation-cutback              0
education-spending                        0
superfund-right-to-sue                    0
crime                                     0
duty-free-exports                         0
export-administration-act-south-africa    0
dtype: int64

In [ ]:
df = df.replace({'?': np.NaN, 'y': 1, 'n': 0})

df.head()

Class Name  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [ ]:
df.describe()

handicapped  ...  export-administration-act-south-africa
count   423.000000  ...                              331.000000
mean      0.442080  ...                                0.812689
std       0.497222  ...                                0.390752
min       0.000000  ...                                0.000000
25%       0.000000  ...                                1.000000
50%       0.000000  ...                                1.000000
75%       1.000000  ...                                1.000000
max       1.000000  ...                                1.000000

[8 rows x 16 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

**Null Hypothesis = The Democrats and Republicans support the water-project-cost-sharing bill in equal numbers.**

**Alternate Hypothesis = The Democrats and Republicans support the water-project-cost-sharing bill in unequal numbers.**

In [ ]:
rep = df[df['Class Name'] == 'republican']
dem = df[df['Class Name'] == 'democrat']

In [ ]:
df.head()

Class Name  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [ ]:
df['water-project-cost-sharing'].value_counts()

1.0    195
0.0    192
Name: water-project-cost-sharing, dtype: int64

In [ ]:
rep = df[df['Class Name'] == 'republican']
dem = df[df['Class Name'] == 'democrat']

In [ ]:
stats.ttest_ind(dem['water-project-cost-sharing'], rep['water-project-cost-sharing'], nan_policy='omit')

#stats.ttest_ind(dem['immigration'], rep['immigration'], nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)


Based on a t-statistic of -0.089, and a p-value of .929, I  do not reject the null hypothesis that republican levels of support are different from democrat levels of support for the water-project-cost-sharing bill.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE